pubsub/kafkapubsub/kafka.go

```diff
@@ -407,75 +407,109 @@ func (s *subscription) Cleanup(sarama.ConsumerGroupSession) error {
-		// s.ch has no buffer, so the send blocks until ReceiveBatch is
-		// ready to receive them.
-		select {
-		case s.ch <- msg:
-		case <-sess.Context().Done():
-			// This session is over, we must return. We'll end up dropping msg, but
-			// that's OK.
-			// TODO(rvangent): See if we can avoid that.
-			// TODO(rvangent): Also consider setting ReceiveBatch to maxMessages=1
-			//                 and maxHandlers=1 similar to NATS.
-		}
-	}
+	s.mu.Lock()
+	s.claims = append(s.claims, claim)
+	s.mu.Unlock()
+	<-sess.Context().Done()
 	return nil
 }
 
 // ReceiveBatch implements driver.Subscription.ReceiveBatch.
 func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
 	// Try to read maxMessages for up to 100ms before giving up.
-	waitForMaxCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	maxWaitCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
 	defer cancel()
 
-	var dms []*driver.Message
 	for {
-		select {
-		case msg := <-s.ch:
-			// Read the metadata from msg.Headers.
-			md := map[string]string{}
-			for _, h := range msg.Headers {
-				md[string(h.Key)] = string(h.Value)
-			}
-			// Add a metadata entry for the message key if appropriate.
-			if len(msg.Key) > 0 && s.opts.KeyName != "" {
-				md[s.opts.KeyName] = string(msg.Key)
-			}
-			ack := &ackInfo{msg: msg}
-			dm := &driver.Message{
-				Body:     msg.Value,
-				Metadata: md,
-				AckID:    ack,
-				AsFunc: func(i interface{}) bool {
-					// TODO(rvangent): Implement.
-					return false
-				},
+		// We'll give up after maxWaitCtx is Done, or if s.closeCh is closed.
+		// Otherwise, we want to pull a message from one of the channels in the
+		// claim(s) we've been given.
+		//
+		// Note: we could multiplex this by ranging over each claim.Messages(),
+		// writing the messages to a single ch, and then reading from that ch
+		// here. However, this results in us reading messages from Kafka and
+		// essentially queueing them here; when the session is closed for whatever
+		// reason, those messages are lost, which may or may not be an issue
+		// depending on the Kafka configuration being used.
+		//
+		// It seems safer to use reflect.Select to explicitly only get a single
+		// message at a time, and hand it directly to the user.
+		s.mu.Lock()
+		cases := make([]reflect.SelectCase, 0, len(s.claims)+2)
+		// A case for s.closeCh being closed, at index = 0.
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(s.closeCh),
+		})
+		// A case for maxWaitCtx being Done, at index = 1.
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(maxWaitCtx.Done()),
+		})
+		// A case per claim.
+		for _, claim := range s.claims {
+			cases = append(cases, reflect.SelectCase{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(claim.Messages()),
+			})
+		}
+		s.mu.Unlock()
+		i, v, ok := reflect.Select(cases)
```

Comment:
`reflect.Select` is rarely used, so sprinkle a few comments in.

Reply:
Can you suggest some? I did sprinkle a few comments in already, and if they're not helping it would help me to know what would help. I agree though, this was new to me.

Reply:
I tried adding some.